lib/kubecfg.libsonnet

```diff
@@ -56,4 +56,55 @@
   // to refer to submatches.  Regex is as implemented in golang regexp
   // package (python-ish).
   regexSubst:: std.native("regexSubst"),
+
+  // isK8sObject(o): Return true iff o is a Kubernetes object.
+  isK8sObject(o):: (
+    std.isObject(o) &&
+    std.objectHas(o, "apiVersion") &&
+    std.objectHas(o, "kind")
+  ),
+
+  // Private helper function for map/fold.
+  local isK8sList(o) = $.isK8sObject(o) && o.apiVersion == "v1" && o.kind == "List",
+
+  // deepMap(func, o): Apply the given function to each Kubernetes
+  // object in nested collection o, preserving the structure of o.
+  deepMap(func, o):: (
+    if isK8sList(o) then
+      o + {items: [func(item) for item in super.items]}
+    else if $.isK8sObject(o) then
+      func(o)
+    else if std.isObject(o) then
+      {[k]: $.deepMap(func, o[k]) for k in std.objectFields(o)}
+    else if std.isArray(o) then
+      [$.deepMap(func, elem) for elem in o]
+    else if o == null then
+      null
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // fold(func, o, init): Apply the given function to each Kubernetes
+  // object in nested collection o, accumulating a result as we go.
+  // Function arg is invoked with arguments (accumulator, object).
+  fold(func, o, init):: (
+    if isK8sList(o) then
+      $.fold(func, o.items, init)
+    else if $.isK8sObject(o) then
+      func(init, o)
+    else if std.isObject(o) then
+      $.fold(func, [o[k] for k in std.objectFields(o)], init)
+    else if std.isArray(o) then
+      std.foldl(function (running, elem) $.fold(func, elem, running), o, init)
+    else if o == null then
+      init
+    else
+      error ("o must be an object or array of k8s objects, found " + std.type(o))
+  ),
+
+  // byGvkName(accum, o): Helper for 'fold'.  This accumulates a
+  // two-level collection of objects by 'apiVersion.kind'
+  // (GroupVersionKind) and then object 'name'.
+  byGvkName(accum, o)::
```

Comment:
should we group such functions in a container like: ?

```
foldFormats: {
  byGvkName(accum, o):: ...
}
```

Reply:
I'm happy to do so if you want, but I think that just increases the typing boilerplate every time you want to use it.

Reply:
`kubecfg.layouts.gvkName`

`kubecfg.layouts.nsGvkName`

Yeah, I think paying a 6 chars price for a bit of self-documentation is worth it

Reply:
Ack, will change.   Just to be clear, an anticipated common case is now going to look like this:

```jsonnet
local objs = kubecfg.fold(kubecfg.layouts.gvkName, kubecfg.renderHelmChart(...), {});
```

One of the original objections to `fold(func, ..)` was the increased boilerplate from having to reference both fold and func (vs some hypothetical combined `toGvkName(input)` that did both in one go).  I'm surprised we're now eager to add extra characters here, which is why I'm resisting.

Reply:
my main objection was not the number of chars per se, but whether it looks easy to understand to a casual user. Compare:

```
local objs = kubecfg.fold(kubecfg.layouts.gvkName, kubecfg.renderHelmChart(...), {});
```

with something like:

```
local objs = kubecfg.renderHelmChart(..., outputLayout=kubecfg.layouts.gvkName);
```

that said; we agreed to not go that route ^^^ and to keep the fold step explicit.